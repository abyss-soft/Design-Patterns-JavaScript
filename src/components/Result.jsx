import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { getAnswers } from '../selectors';

const StyledResult = styled.p`
  color: ${props => props.theme.text};
  margin: 3rem 0;
  text-align: center;
`;

export const Result = ({ answers }) => {
  let correct = 0;

  answers.map(answer => (answer.correct ? correct++ : null));

  return (
    <StyledResult>
      You got <strong>{correct}</strong> patterns right out of {answers.length}.
    </StyledResult>
  );
};

Result.propTypes = {
  answers: PropTypes.array.isRequired
};

export default connect(state => ({
  answers: getAnswers(state)
}))(Result);
