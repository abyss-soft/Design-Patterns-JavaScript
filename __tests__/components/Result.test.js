import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import 'jest-styled-components';
import Result from '../../src/components/Result';

const mockStore = configureMockStore();
const store = mockStore({
  progress: {
    answers: [
      { id: 1, correct: true },
      { id: 2, correct: true },
      { id: 3, correct: true },
      { id: 4, correct: false }
    ]
  }
});

describe('<Result /> component', () => {
  let result;

  beforeEach(() => {
    // the wrapper DIV added to get rid of iterator key warning
    result = shallow(
      <div key="key">
        <Provider store={store} key="unique">
          <Result key="result" />
        </Provider>
      </div>
    );
  });

  it('renders without crashing', () => {
    expect(result).toBeTruthy();
  });

  it('shows the number of correct answers', () => {
    const expected = '<strong>3</strong> patterns right out of 4.';
    expect(result.html()).toEqual(expect.stringContaining(expected));
  });

  it('has styled-component rendered classes', () => {
    const tree = renderer.create(result).toJSON();
    expect(tree.children[0].props.className).toBeDefined();
  });

  it('matches the snapshot', () => {
    const tree = renderer.create(result).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
