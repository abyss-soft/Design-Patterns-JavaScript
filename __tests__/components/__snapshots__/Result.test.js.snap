// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Result /> component matches the snapshot 1`] = `
.c0 {
  margin: 3rem 0;
  text-align: center;
}

<div>
  <p
    className="c0"
  >
    You got 
    <strong>
      3
    </strong>
     patterns right out of 
    4
    .
  </p>
</div>
`;
